rfcs/013-gitlab-spam-checker/013-gitlab-spam-checker.md

```diff
@@ -0,0 +1,93 @@
+
+Similarly to GitLab's Spamcheck service, storing the model in an S3 bucket and pulling it to the spam management service on start-up would allow the model to be updated without having to redeploy the spam management service.
+
+An allowlist of trusted email domain providers could be implemented to prevent false positives to ensure users with trusted email domains can post content without it being checked for spam. YAT already possesses a list of trusted email domains.
+
+This mermaid diagram provides an overview of how this system would work:
+
+```mermaid
+  graph TB;
+    GH["GitLab Hooks"] -- "Triggers on Create/Update" --> SC["Spam Checking Service"]
+    AR["Abuse Reports"] -- "Retrieved from GitLab" --> SC
+    SC -- "Checks" --> I["Issues"]
+    SC -- "Checks" --> N["Notes"]
+    SC -- "Checks" --> P["Projects"]
+    SC -- "Checks" --> U["Users"]
+    S3ML["ML model stored in S3 Bucket"] -- "On Start-Up" --> SC
+    TED["Trusted Email Domains List"] -- "Skips Spam Checks" --> SC
+    SC -- "Invokes" --> ML["Machine Learning Model"]
+    ML -- "Classifies" --> SP["Spam"]
+    SP -- "Identified" --> SC
+    SC -- "Deletes via API" --> GD["GitLab"]
+    SJ["Snippet Job"] -. "Runs Periodically" .-> SC
+    SC -- "Logs Decisions & Confidence Levels" --> L["Logs"]
+    SC -- "Stores Evaluated Content" --> S3["S3 Bucket"]
+    linkStyle 0 stroke:#2ecd71,stroke-width:2px;
+    linkStyle 1 stroke:#2ecd71,stroke-width:2px;
+    linkStyle 2 stroke:#2ecd71,stroke-width:2px;
+    linkStyle 3 stroke:#2ecd71,stroke-width:2px;
+    linkStyle 4 stroke:#2ecd71,stroke-width:2px;
+    linkStyle 5 stroke:#2ecd71,stroke-width:2px;
+    linkStyle 6 stroke:#2ecd71,stroke-width:2px;
+    linkStyle 7 stroke:#2ecd71,stroke-width:2px;
+    linkStyle 8 stroke:#2ecd71,stroke-width:2px;
+    linkStyle 9 stroke:#2ecd71,stroke-width:2px,dasharray: 5, 5;
+    linkStyle 10 stroke:#2ecd71,stroke-width:2px;
+    linkStyle 11 stroke:#2ecd71,stroke-width:2px;
+    linkStyle 12 stroke:#2ecd71,stroke-width:2px;
+```
+
+## Drawbacks
+
+1. Ensuring the system can handle many system hooks from GitLab without significant latency is critical. It needs to be scalable and able to handle peak loads.
+2. The spam management service will require a GitLab personal access token with administrative rights to delete spam.
+3. We must ensure we can be informed of and restore incorrectly classified content.
+
+## Rationale and Alternatives
+
+This solution is deemed the best as it provides real-time spam detection without intruding on the user experience. Alternatives such as manual monitoring are too time-consuming.
+
+Not implementing this solution would mean continuing to expose GitLab users to potential spam, which could disrupt their experience and pose security risks, and open RenkuLab.io up to promoting unethical content.
+
+## Unresolved Questions
```

Comment:
How to calibrate it such that it can be trusted to act sensibly on obvious spam, and to flag possible spam to administrators without having too many false positives...

Reply:
Yeah, we'll take that up with innovation team because I think that will largely depend on the specifics of the model